### ISSUE TITLE
AR: Weird query when chaining `uniq` and `count`: excessive `DISTINCT`

### ISSUE THREAD

**Issue by waterlink** [rails/rails#18694]
Consider this statement:

``` ruby
User.uniq.count    # it can and often do have some `where`-s and `join`-s before `uniq`, but this statement already shows the problem
```

generates this SQL:

``` sql
SELECT DISTINCT COUNT(DISTINCT `users`.`id`) FROM `users`
```

where one would expect having just:

``` sql
SELECT COUNT(DISTINCT `users`.`id`) FROM `users`
```

(no redundant `DISTINCT`). Usually it does not make a difference, since `DISTINCT` out of one value (result of count) is the same, but if you look at MySQL docs: http://dev.mysql.com/doc/refman/5.5/en/group-by-optimization.html it says that it will not use loose index scan with this query. Which actually makes query extremely slow on big (millions) table(s). Removing excessive `DISTINCT` makes it use right indexes and it becomes very fast.

Is it expected behavior? Is there any workaround except avoiding `uniq` + `count` ?

Rails: 4.0.12
Ruby: 2.1.5
MySQL: Percona 5.6

**Comment by sgrif** [comment]
Would you please provide a script that we can use to reproduce the issue, using this template? Additionally, Rails 4.0 is no longer supported. This issue must be reproducible on 4.2 or master.

**Comment by waterlink** [comment]
@sgrif thank you for your answer, I will prepare an example soon for that one.

**Comment by habermann24** [comment]
The same applies for Rails 4.2.
Not sure if this is really a bug though?

```ruby
# Activate the gem you are reporting the issue against.
gem 'activerecord', '4.2.0'
require 'active_record'
require 'minitest/autorun'
require 'logger'

# Ensure backward compatibility with Minitest 4
Minitest::Test = MiniTest::Unit::TestCase unless defined?(Minitest::Test)

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: 'sqlite3', database: ':memory:')
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :users, force: true do |t|
  end
end

class User < ActiveRecord::Base
end

class BugTest < Minitest::Test
  def test_uniq_count
    stream = capture(:stdout) do
      User.uniq.count
    end

    assert_equal 1, stream.scan(/DISTINCT/).count
  end
end
```

**Comment by hbd225** [comment]
Sorry I didn't realize there was already a PR.

Should we fix arel rather than relation?
I found arel added distinct statement twice at following code.

https://github.com/rails/arel/blob/master/lib/arel/visitors/to_sql.rb#L229

**Comment by ClaudioFloreani** [comment]
Which Rails version is supposed to fix this issue? It's still there in Rails 4.2.2:

    2.2.1 :144 > Rails.version
     => "4.2.2" 
    2.2.1 :145 > Box.distinct.count
       (128.5ms)  SELECT DISTINCT COUNT(DISTINCT "boxes"."id") FROM "boxes"
     => 7

**Comment by ClaudioFloreani** [comment]
I think this is only the tip of a more extended issue: https://github.com/rails/rails/issues/21799

**Comment by kamipo** [comment]
Fixed in #27615.
